src/support/rand.c

```diff
@@ -28,43 +28,68 @@
 
 #include "wt_internal.h"
 
+/*
+ * This is an implementation of George Marsaglia's multiply-with-carry pseudo-
+ * random number generator.  Computationally fast, with reasonable randomness
+ * properties.
+ *
+ * Be very careful about races here. Multiple threads can call __wt_random
+ * concurrently, and it is okay if those concurrent calls get the same return
+ * value. What is *not* okay is if reading/writing the shared state races and
+ * uses two different values for m_w or m_z. That can result in a stored value
+ * of zero, in which case they will be stuck on zero forever. Take a local copy
+ * of the values to avoid that, and read/write in atomic, 8B chunks.
+ */
 #undef	M_W
-#define	M_W	(rnd)[0]
+#define	M_W(p)	((uint32_t *)&(p))[0]
 #undef	M_Z
-#define	M_Z	(rnd)[1]
+#define	M_Z(p)	((uint32_t *)&(p))[1]
 
 /*
  * __wt_random_init --
  *	Initialize return of a 32-bit pseudo-random number.
  */
 void
-__wt_random_init(uint32_t *rnd)
+__wt_random_init(uint64_t * volatile rnd_state)
 {
-	M_W = 521288629;
-	M_Z = 362436069;
+	uint64_t rnd;
+
+	M_W(rnd) = 521288629;
+	M_Z(rnd) = 362436069;
+	*rnd_state = rnd;
 }
 
 /*
  * __wt_random --
  *	Return a 32-bit pseudo-random number.
- *
- * This is an implementation of George Marsaglia's multiply-with-carry pseudo-
- * random number generator.  Computationally fast, with reasonable randomness
- * properties.
- *
- * We have to be very careful about races here.  Multiple threads can call
- * __wt_random concurrently, and it is okay if those concurrent calls get the
- * same return value.  What is *not* okay is if reading the shared state races
- * with an update and uses two different values for m_w or m_z.  That could
- * result in a value of zero, in which case they would be stuck on zero
- * forever.  Take local copies of the shared values to avoid this.
  */
 uint32_t
-__wt_random(uint32_t *rnd)
+__wt_random(uint64_t * volatile rnd_state)
 {
-	uint32_t w = M_W, z = M_Z;
+	uint64_t rnd;
+	uint32_t w, z;
+
+	/*
+	 * Take a copy of the random state so we can ensure that the
+	 * calculation operates on the state consistently regardless of
+	 * concurrent calls with the same random state.
+	 */
+	rnd = *rnd_state;
+	w = M_W(rnd);
+	z = M_Z(rnd);
+
+	/*
+	 * I don't know the period of this PRNG. Do a cheap check, if the value
+	 * goes to 0 (from which we won't recover), reset to the initial state.
+	 * This has additional value if a caller fails to initialize the state,
+	 * or initializes with a seed that results in a short period.
+	 */
+	if (z == 0 || w == 0)
```

Comment:
I think it's necessary: I'm checking on this, but even if the period si as big as we think it is (2^60), it doesn't hurt to do a check of values in a register, does it?